### PR TITLE
feat(Dockerfile): Run as nonroot user

### PR DESCRIPTION
**Pull request by jchauncey**
This pr moves the final image from scratch to distroless/static:nonroot

### CODE REVIEW

**Comment by jchauncey** [comment]
still need to test that this actually works

**Comment by jeefy** [comment]
Unfortunately this won't work with scratch because of the file permissions with tmp (just verified)

That said, if you change the base image from "scratch" to "distroless", this should work:

```
FROM gcr.io/distroless/base AS final
```

I think that change on top of your user change will shore things up. :)

**Comment by jchauncey** [comment]
yeah i tried starting it locally and noticed the tmp error. We could just use distroless:nonroot

**Comment by jchauncey** [comment]
Ok i updated to use nonroot distroless which seems it may have fixed the tmp issue.

File: Dockerfile [hunk 1]
```diff
@@ -31,7 +31,7 @@ RUN echo "Building for $GOARCH" \
     && GOARCH=${GOARCH} hack/build.sh 
 
 # Final stage: the running container.
-FROM scratch AS final
+FROM gcr.io/distroless/static:nonroot AS final
 
 # Import the Certificate-Authority certificates for enabling HTTPS.
 COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
```
Comment (sozercan) [inline Dockerfile]: ca-certificates are included in distroless already, no need to copy from builder

**Comment by jchauncey** [comment]
fixes #16

**Comment by jeefy** [comment]
This is great, thanks for tackling this!

/approve
/assign @maciaszczykm @floreks

**Comment by floreks** [comment]
Is there any actual reason to use distroless instead of scratch? The container itself does not enforce privileges. We do restrict container privileges in our deployment files already.

https://github.com/kubernetes/dashboard/blob/5a289e51a60958bafacb5266dc602ca4bb565806/aio/deploy/recommended.yaml#L288-L292

**Comment by jchauncey** [comment]
Our customers (aks) are having this clusters scanned with the twistlock tool and its flagging this image (along with a few others) as being created with a nonroot user. Even though you set the user at runtime that isnt picked up by the scanning tool and I believe this is failing the scan as its marked a high severity.

**Comment by jeefy** [comment]
@floreks Distroless is widely used elsewhere across K8s and solves a couple
problems that we can run into with scratch (certificates, nonroot, and tmp
are the three primary ones)

It's a small change for a pretty big win IMO and strongly endorse the move.

**Comment by floreks** [comment]
Why I like scratch is because it does not add any additional layers to the final image. It is basically an empty image with just the binary. Distroless images have all this additional stuff that we do not need and will most probably never need. If some static check fails because it cannot see nonroot user inside dockerfile then we can simply add `passwd` file and change user without changing the base image. We've already had this discussion inside dashboard repository https://github.com/kubernetes/dashboard/pull/4301.

**Comment by sozercan** [comment]
distroless/static only have these components: https://github.com/GoogleContainerTools/distroless/blob/master/base/README.md

- ca-certificates and /tmp are already being used here
- as you mentioned, /etc/passwd for adding (nonroot) user so the container itself doesn't get flagged

**Comment by floreks** [comment]
We don't really need `ca-certificates` or `tmp` dir already inside the image. I'd only add `passwd` to the current Dockerfile, then change user and leave it as is.

**Comment by jchauncey** [comment]
I must be doing something wrong then because I cant seem to get that to work. It complains about writing to /tmp.

**Comment by floreks** [comment]
tmp dir is handled by our yaml deployment file.

**Comment by jchauncey** [comment]
Ok new commit that removes /tmp and ca-certificates as well as creates a user in the builder stage and copies it to the final image. went ahead and pushed this commit to see if people are ok with this approach. still need to verify it works

**Comment by jchauncey** [comment]
Ok i verified that the pod starts and runs with no errors in the logs with the latest commit

**Pull request merged**